### DISCUSSION TITLE
Rely more on Mule's configuration when truncating author names in commit logs

### DISCUSSION THREAD

**Discussion by nbarrientos** [magit/magit#4907] [Ideas]
Hi,

It'd be nice if the ellipsis used to truncate strings representing author names in `magit-log` buffers was customisable. If Magit relied on `truncate-string-to-width`'s default (which, at least in Emacs 28, defaults to calling `truncate-string-ellipsis`) it would permit Magit to follow system-wide customisations done by modifying the variable `truncate-string-ellipsis`.

`truncate-string-ellipsis` [takes care](https://git.savannah.gnu.org/cgit/emacs.git/tree/lisp/international/mule-util.el?h=emacs-28#n52) of making sure that, if `truncate-string-ellipsis` is not customised, the default ellipsis char (`"…"`) is only displayed if `char-displayable-p` is `t` (similar to what Magit is hardcoding now). Something I don't know is why it was chosen in the past to use `">"` if `"…"` cannot be displayed (part of https://github.com/magit/magit/commit/ce34a59746baf0d94f0b7ed157b6e740924ce64a).

If you don't mind falling back to Mule's default (`"..."`) when `"…"` cannot be displayed in this case, it'd be cool if the attached patch was considered to, as mentioned above, allow users to customise ellipses when displaying commit logs using the "standard" mechanism provided by Mule.

Happy to submit a merge request.

Thanks.

```patch
From 4d987f4fc2cfa8fe317e991b2bf4254c6cd12e48 Mon Sep 17 00:00:00 2001
From: Nacho Barrientos <nacho.barrientos@cern.ch>
Date: Wed, 22 Mar 2023 20:14:36 +0100
Subject: [PATCH] magit-log: Rely on Mule's configuration to display ellipses

---
 lisp/magit-base.el | 2 +-
 lisp/magit-log.el  | 3 +--
 2 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/lisp/magit-base.el b/lisp/magit-base.el
index ace31fca..f4e89f45 100644
--- a/lisp/magit-base.el
+++ b/lisp/magit-base.el
@@ -392,7 +392,7 @@ defcustom magit-ellipsis
 as well, but not the following:
 
 - Author names in the log margin are always abbreviated using
-  \"…\" or if that is not displayable, then \">\".
+  `truncate-string-ellipsis'.
 
 - Whether collapsed sections are indicated using ellipsis is
   controlled by `magit-section-visibility-indicator'."
diff --git a/lisp/magit-log.el b/lisp/magit-log.el
index fb5b844e..7d124573 100644
--- a/lisp/magit-log.el
+++ b/lisp/magit-log.el
@@ -1530,8 +1530,7 @@ defun magit-log-format-author-margin
                            (truncate-string-to-width
                             (or author "")
                             details-width
-                            nil ?\s
-                            (if (char-displayable-p ?…) "…" ">"))
+                            nil ?\s t)
                            'magit-log-author)
                           " "))
              (magit--propertize-face
-- 
2.40.0
```

**Comment by nbarrientos** [comment]
Ok, nevermind, I understand this is far from being the most exciting patch to look at :) I'll just carry it along locally in my Emacs config.

I think that it'd be useful to allow the users to control the ellipsis as some fonts' implementation of the glyph might be too wide, breaking the nice column alignment in some cases, e.g.:

![image](https://user-images.githubusercontent.com/131893/228815882-89eecb10-a5d3-4335-9759-59fefde88e02.png)

vs

![image](https://user-images.githubusercontent.com/131893/228815680-8944310d-27d7-4795-8171-45ab352878dd.png)

**Reply by tarsius** [reply]
I did have a quick look at the relevant code when you opened this, but I did not get around to replying until now, because I am currently trying not to jump on every issue as it comes up.

**Comment by tarsius** [comment]
> Something I don't know is why it was chosen in the past to use `">"` if `"…"` cannot be displayed (part of https://github.com/magit/magit/commit/ce34a59746baf0d94f0b7ed157b6e740924ce64a).

Because it is only one instead of three characters long, and since it is being used in a rather narrow column, that does matter.

> It'd be nice if the ellipsis used ... it would permit Magit to follow system-wide customisations done by modifying the variable `truncate-string-ellipsis`.

IMO the default global values work well in places where something might have to be truncated occasionally. In this column, in which well over 50% of cells might have to be truncated, all the time, the needs are a bit different.

> if `truncate-string-ellipsis` is not customised, the default ellipsis char (`"…"`) is only displayed if `char-displayable-p` is `t` (similar to what Magit is hardcoding now).

However if it is customized, then the same is displayed in gui and cli, except if `char-displayable-p` returns nil for the chosen string, in which case the hardcoded `...` is used. In other words, it is not possible to pick a custom value to be used *only* in cli.

Wanting to use something other than ">" or "..." in the terminal is reasonable, but so is not wanting to sacrifice using `…` in gui.

In short, `truncate-string-ellipsis` is not suitable and we have to provide another option whose value is a cons-cell. I.e., we have to add `magit-log-margin-ellipsis` in addition to the existing `magit-ellipsis`. I do not want to use `magit-ellipsis` here because I think `...` works well in most places and where that is the case it is preferable to `>`, but in the log margin something that is just one character long is preferable.

**Reply by nbarrientos** [reply]
Thanks Jonas for your input.

I understand. Instead of adding yet another customisation option as you're suggesting, would you be open to giving more responsibilities to `magit-ellipsis` at the expense of breaking backwards compatibility? Something along the lines of https://github.com/nbarrientos/magit/commit/9b849e7d8359d2e02803beb44834a18806fe0626. It's a bigger patch but it leaves a clean and extensible code base in case another ellipsis-related casuistic showed up in the future.

**Reply by tarsius** [reply]
That looks good. I do somewhat doubt that a need for a third style will arise, but since this doesn't have a larger surface area than adding a second option, and it might turn out I was wrong about no third style being needed, I think going with this approach is the right move.

Please open a pull-request. I do have some minor requests but its easier to discuss those once there is a pr.

**Reply by nbarrientos** [reply]
Great! Here it is: https://github.com/magit/magit/pull/4912

Feel free to close this discussion :smiley:
